### PR TITLE
Addresses Regex Perf Issue 32764

### PR DESCRIPTION
**Pull request by Alois-xx**
This fix should save ca 40% for the common case when many Char.ToLower calls are emitted which access CultureInfo.CurrentCulture for compiled Regex Queries.

### CODE REVIEW

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs [hunk 1]
```diff
@@ -961,14 +962,20 @@ private void Advance()
...
             else
                 Call(s_getCurrentCulture);
 
-            Call(s_chartolowerM);
+            Stloc(_cultureV);
+        }
+
+        private void CallToLower()
+        {
+             Ldloc(_cultureV);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs]: What is the lifetime of this?  e.g. does this cache end up spanning multiple regex calls and user code such that user code could change the current culture and end up with different behavior than before?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs [hunk 2]
```diff
@@ -961,14 +962,20 @@ private void Advance()
...
+             Call(s_chartolowerM);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs]: @ViktorHofer, am I misremembering, or did you do some ToLower-related optimizations/changes when you were cleaning up the regex code recently?  If yes, should we instead just port the appropriate changes to the compiled version?  If no, is there a similar optimization to be done on the interpreted side of things?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs (hunk 1 above)
Comment (Alois-xx) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs]: _cultureV is a local variable. The lifetime of the cache is therefore method local in FindFirstChar and Go methods of the compiled regular expression.

Comment (stephentoub) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs]: Ah, I see. Thanks.

**Comment by stephentoub** [comment]
Do we have regex tests (interpreted and compiled) that rely on the current culture being something specific? If not, it'd be good to add as part of this.

**Comment by ViktorHofer** [comment]
> Do we have regex tests (interpreted and compiled) that rely on the current culture being something specific? If not, it'd be good to add as part of this.

Only very few for Unicode characters. We should definitely add a bunch.

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs (hunk 2 above)
Comment (ViktorHofer) [reply src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs]: Right, but only minor allocation optimizations. I wouldn't do that as part of this PR but revisit the compiled code paths later to make sure that we don't diverge and bring optimizations over.

**Comment by Alois-xx** [comment]
Benchmark.NET was a little funny to get it to use the compiled dlls and not the ones from the framework sdk. Here are the patched Results:

```
                        Method | N |         Mean |        Error |       StdDev | Allocated |
------------------------------ |-- |-------------:|-------------:|-------------:|----------:|
                      Original | 1 | 438,212.4 ns | 4,647.930 ns | 4,347.676 ns |       0 B |
                 CaseSensitive | 1 | 227,563.0 ns | 5,142.693 ns | 5,281.167 ns |       0 B |
 CaseSensitiveCultureInvariant | 1 | 215,541.1 ns | 1,187.442 ns | 1,110.734 ns |       0 B |
          SimplifiedExpression | 1 |   1,525.5 ns |    11.969 ns |    11.195 ns |       0 B |
      Case_Sensitive_Substring | 1 |     691.5 ns |     4.897 ns |     4.581 ns |       0 B |
```
And here the baseline version where I did revert the changes of my commit to really measure the impact of only my change:
```
                        Method | N |           Mean |         Error |        StdDev | Allocated |
------------------------------ |-- |---------------:|--------------:|--------------:|----------:|
                      Original | 1 | 1,055,537.2 ns | 31,271.169 ns | 36,011.922 ns |       0 B |
                 CaseSensitive | 1 |   216,579.6 ns |  1,827.153 ns |  1,709.119 ns |       0 B |
 CaseSensitiveCultureInvariant | 1 |   217,116.7 ns |  2,212.376 ns |  1,961.213 ns |       0 B |
          SimplifiedExpression | 1 |     1,523.0 ns |     13.191 ns |     12.339 ns |       0 B |
      Case_Sensitive_Substring | 1 |       671.3 ns |      8.552 ns |      8.000 ns |       0 B |
```
Just in case you did not see here is the ETW chart:
grafik

Actually the Benchmark.NET numbers are even better. From the numbers it looks like the slow code gen path is only triggerd when RegexOptions.IgnoreCase is used. But anyway this is certainly a widely used option.

**Comment by Alois-xx** [comment]
@stephentoub: I could think of some Turkish I tests which behave differently in different locales. That would be a good test to verify that the right culture was used.

**Comment by ViktorHofer** [comment]
That sounds good and should suffice. As mentioned before, we currently don't have a comprehensive set of inputs with different cultures.

**Comment by Alois-xx** [comment]
@ViktorHofer: Added some locale tests which really show different behavior under the turkish locale.

**Comment by Alois-xx** [comment]
@dotnet-bot test this please

File: src/System.Text.RegularExpressions/tests/RegexCultureTests.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        /// See https://en.wikipedia.org/wiki/Dotted_and_dotless_I 
+        /// </summary>
+        [Fact]
+        public void TurkishI_Is_Differently_LowerUpperCased_In_Turkish_Culture()
+        {
+            var turkish = new CultureInfo("tr-TR");
+            string input = "Iıİi";
+
+            var cultInvariantRegex = Create(input, CultureInfo.InvariantCulture, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
+            var turkishRegex =       Create(input, turkish,                      RegexOptions.IgnoreCase);
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/tests/RegexCultureTests.cs]: Nit: please change var to be the actual type here... we only use var when the type is obvious from the right-hand side, namely a ctor or an explicit cast.

**Comment by Alois-xx** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
Thanks!

**Comment by danmosemsft** [comment]
Nice win here @alois-xx many thanks! Any interest in more work on regex?

**Comment by Alois-xx** [comment]
@danmosemsft: If I find new issues or ideas I will definitely file an issue. But currently my free time is quite limited.

**Pull request merged**